Msilva/UIE-93 decouple data-clients - useWorkspaces

msilva-broad opened the pull request:
### Jira Ticket: https://broadworkbench.atlassian.net/browse/UIE-93

## Summary of changes:
To better support composability of Environments Page component, useWorkspaces hook is now made up of more composable patterns.

### What
- code cleanup and refactor.  should have no functional change.
- useWorkspaces hook is now moved over to src/workspaces feature area, and is now made of a composable inner hook to aid in code reuse.
- useWorkspaces adds terra-ui specific data-call and concerns which AoU or other apps can chose to compose differently in their implementation.
- added useAutoLoadedData to streamline cases where initial data load is desired.
- added useCachedData to streamline cases where data should be cached in a state atom.
- added useLoadedDataEvents to streamline optional use of onError and/or onSuccess triggering from useLoadedData family of hooks.
- added WorkspaceProvider to put a layer around wonky Ajax data patterns. 
- added useSettableStore utility hook to offer useStore interface that is more like useState.  

### Why
- To better support composability of Environments Page component usage in other apps like AoU

### Testing strategy
- [x] basic smoke test
- [ ] more robust testing

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- moved useWorkspaces hook out of worspaces-utils and into new src/worspaces functional area.
- converted useWorkspaces to Typescript
- implemented WorkspacesProvider to break out the ajax data needs of useWorkspaces.
- made useWorkpaces composable.
- added unit test coverage.
- TODO: solve for useReplaceableAjaxExperimental hook should not be called from provider.

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- useWorkspaces hook is now moved over to src/workspaces feature area, and is now composable.
- dependencies for data-provider and state container can be given, but terra-ui expected composition is also exported as the primary useWorkspaces export to not break consumers.
- added useSettableStore utility hook to offer useStore interface that is mor like useState.
- TODO: still need to make error notification composable as well.

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- fixed one more useWorkspaces reference that slipped in after rebase

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- attempting unit test fix failing in PR

nawatts commented on src/libs/ajax/workspaces/providers/WorkspaceProvider.test.ts:
```diff
@@ -0,0 +1,52 @@
+
+/**
+ * local test utility - mocks the Ajax super-object and the subset of needed multi-contracts it
+ * returns with as much type-saftely as possible.
```
Nit: spelling.

```suggestion
 * returns with as much type-safety as possible.
```

msilva-broad replied:
fixed

nawatts commented on src/libs/ajax/workspaces/providers/WorkspaceProvider.test.ts:
```diff
@@ -0,0 +1,52 @@
+type WorkspacesAjaxContract = Pick<AjaxContract, 'Workspaces'>['Workspaces'];
+type WorkspacesAjaxNeeds = Pick<WorkspacesAjaxContract, 'list'>;
+
+interface AjaxMockNeeds {
```
Is there an advantage to creating all these types vs typing a mocked Ajax return value with `DeepPartial<AjaxContract>`?

msilva-broad replied:
the object/function layering makes it gnarly.... and because I want to assert on call arguments to some of those "mid-functions, this helper (fixture) pattern is working well to remove that repetition from the // Arrange sections of the tests.

nawatts replied:
Ok. Yeah, after dealing with DataRepo's Ajax module, this makes more sense to me.

nawatts commented on src/libs/ajax/workspaces/providers/WorkspaceProvider.ts:
```diff
@@ -0,0 +1,22 @@
+
+export type FieldsArg = string[];
+export type WorkspaceListOptions = AbortOption & {
+  stringAttributeMaxLength?: string | number;
```
The API types this parameter as an integer.
https://rawls.dsde-prod.broadinstitute.org/#/workspaces/listWorkspaces

```suggestion
  stringAttributeMaxLength?: number;
```

msilva-broad replied:
yes, but it appears to be passed in as a string value in the attribute request header.... number will format fine, so I chose the union of both.

nawatts replied:
Sure, everything gets formatted as a string when it's encoded in the URL. But logically, it should be a number so we should use TS to enforce that only a number is provided. Accepting a string would also allow values like `stringAttributeMaxLength: 'notANumber'`.

msilva-broad replied:
ok yeah, great point.  will change

msilva-broad replied:
fixed

nawatts commented on src/libs/ajax/workspaces/providers/WorkspaceProvider.ts:
```diff
@@ -0,0 +1,22 @@
+  list: async (fieldsArgs: FieldsArg, options: WorkspaceListOptions): Promise<WorkspaceWrapper[]> => {
+    const { signal, stringAttributeMaxLength } = options;
+
+    const ws = await Ajax(signal).Workspaces.list(fieldsArgs, stringAttributeMaxLength);
+    return ws;
```
Style nit...

If terseness is a goal, we could return the result of `list` directly.

```suggestion
    return Ajax(signal).Workspaces.list(fieldsArgs, stringAttributeMaxLength);
```

Or if descriptiveness is a goal, then I'd suggest not abbreviating variable names.

```suggestion
    const workspaces = Ajax(signal).Workspaces.list(fieldsArgs, stringAttributeMaxLength);
    return workspaces;
```

msilva-broad replied:
this was unmodified from the original code.... I will go with the 2nd suggestion.

nawatts commented on src/libs/react-utils.ts:
```diff
@@ -182,6 +182,18 @@ export const useStore = <T>(theStore: Atom<T>): T => {
+ * When the store changes, the component will re-render.
+ * @param theStore
+ */
+export const useSettableStore = <T>(theStore: Atom<T>): [T, (value: T) => void] => {
```
I'm not convinced this needs to exist. The component already has to have a reference to the store to call this function, so it could call `store.set` directly.

I'd be in favor of changing `useStore` to this interface, but I think adding this in addition to `useStore` adds more complexity than it's worth.

nawatts commented on src/pages/EnvironmentsPage.test.ts:
```diff
@@ -55,9 +56,9 @@ describe('Environments Page', () => {
-    const watcher = Environments;
-    expect(watcher).toBeCalledTimes(1);
-    expect(watcher).toBeCalledWith(
+    const environmentsWatcher = Environments;
+    expect(environmentsWatcher).toBeCalledTimes(1);
+    expect(environmentsWatcher).toBeCalledWith(
```
Nit: I think it's clearer to leave out the extra variable and assert using the spied/mocked function/component directly.

```suggestion
    expect(Environments).toBeCalledTimes(1);
    expect(Environments).toBeCalledWith(
```

Same comment elsewhere in this file.

msilva-broad replied:
fixed

nawatts commented on src/workspaces/useWorkspaces.composable.ts:
```diff
@@ -0,0 +1,75 @@
+   * or an equivalent like useSettableStore.  Implementation is expected to be a React Hook that will
+   * re-render a consuming component on state update.
+   */
+  useWorkspacesState: () => [WorkspaceWrapper[], (newValue: WorkspaceWrapper[]) => void];
```
I'm not keen on having this injected. I think `useWorkspaces` vs `useWorkspacesState` is a subtle difference and likely to be confusing. I found it difficult to understand when reviewing this PR.

To support Terra UI's global workspaces store, I think I'd propose an optional `onLoad` (or similar name) callback similar to the `onError` callback will have to be added to support error notifications.

Or... Terra UI's `useWorkspaces` could wrap the base `useWorkspaces` and use an effect to update the store.
```ts
const useWorkspaces = () => {
  const result = baseUseWorkspaces();
  useEffect(() => {
    workspacesStore.set(result.workspaces);
  }, [result.workspaces])
  return result
})
```

That would simplify the `makeUseWorkspaces` interface.

msilva-broad replied:
hmmm.... kicking this around a bit locally... um, what about initial value?

A benefit of the proposed useWorkspacesState mechanic is that it provides initial value from the store (atom)... and while we established that this hook is reloading on every initial mount, I am a little hesitant to remove the expectation that there is an initial useful value for a given consumer of the hook.

I am thinking streamlining the composable pattern quite a bit more.... but keep the state mechanic and make it internal

```
export const useWorkspaces: UseWorkspacesState = (
    fieldsArg?: FieldsArg,
    stringAttributeMaxLength?: string | number
): UseWorkspacesStateResult => {
    const useWorkspacesStore = () => useSettableStore(workspacesStore);
    
    return useWorkspacesComposable({
        workspaceProvider,
        useWorkspacesStore,
    }, fieldsArg, stringAttributeMaxLength);
};
```

nawatts replied:
Ideally, I suppose it would be preferable to take an initial value. However, practically, it doesn't matter. The workspaces store is initialized with an empty array and since `makeUseWorkspaces` is run at module import/app load, the initial value will always be an empty array. We'd only have to worry about this if we started caching the workspaces store in local storage, etc.

nawatts replied:
Oh, I misunderstood...

But I still think that returning an initial value from the store can be done without injecting the state into the generic useWorkspaces.

```ts
const useWorkspaces = (...args) => {
  const { workspaces, loading, error } = baseUseWorkspaces(...args)

  useEffect(() => {
    workspacesStore.set(workspaces)
  }, [workspaces])

  const storedWorkspaces = useStore(workspacesStore)
  return { workspaces: storedWorkspaces, loading, error }
})
```

And the same pattern could probably be used for error notifications, allowing the generic useWorkspaces to only require a workspaces provider:

```ts
const useWorkspaces = (...args) => {
  const { workspaces, loading, error } = baseUseWorkspaces(workspacesProvider, ...args)
  
  useEffect(() => {
    workspacesStore.set(workspaces)
  }, [workspaces])

  useEffect(() => {
    if (error) {
       reportError('Error loading workspaces' error)
    }
  }, [error])

  const storedWorkspaces = useStore(workspacesStore)
  return { workspaces: storedWorkspaces, loading, error }
})
```

msilva-broad replied:
hmmm....

msilva-broad replied:
ok, what you suggest will work... as long as useWorkspaces is the only source of truth.... the way the atoms are intended to function is a more centralized store which many parts of the UI can treat as the source of truth.  Looking through the suggestion and the hook internals of useWorkspaces, it would have to assume an empty [] array as initial value on mount... clearing out any previously set value in workspacesStore.

nawatts replied:
> Looking through the suggestion and the hook internals of useWorkspaces, it would have to assume an empty [] array as initial value on mount... clearing out any previously set value in workspacesStore.

Ah, because this would reset the store with the initial workspaces list returned from `baseUseWorkspaces`?

```ts
const { workspaces, loading, error } = baseUseWorkspaces(workspacesProvider, ...args)
  
useEffect(() => {
  workspacesStore.set(workspaces)
}, [workspaces])
```

I think that could fixed by only updating the store when loading is done:

```ts
const { workspaces, loading, error } = baseUseWorkspaces(workspacesProvider, ...args)
  
useEffect(() => {
  if (!loading) {
    workspacesStore.set(workspaces)
  }
}, [workspaces, loading])
```

msilva-broad replied:
nope, useWorkspaces onMount would still clear out central store to [] empty array since loading is false then too.   We would likely have to pass in an initial value, which creates it's own spagetti.... landing back to my current proposal of just providing a store dependency which feels cleaner.

nawatts replied:
I think that's an issue with the initial value of `loading` then. If it immediately starts loading data, why is `loading` not initialized to true?

This is a frustration I've frequently run into with `useLoadedData`: the NoneState is rarely useful and I almost always want the data to immediately start loading.

msilva-broad replied:
baby steps.... I am resisting bringing in useLoadedData to keep the PR risk down.... useLoadedData initial NoneState could actually help here if you want to keep the wrapper-hook-adds-new-concern (store) concern.... but I argue that the store dependency handles this well....

Because if we want to preserve the current assumption that atoms are CENTAL sources of truth, making the store a dependency is cleaner in my opinion then the wrapper cruft required... what about the case where an external actor changes the store?  useLoadedData needs that update to hold up on the current set of assumptions...

nawatts replied:
My objection to the store as dependency is that it leaks Terra UI specific concerns into shared, preferably generic, code.

> what about the case where an external actor changes the store? useLoadedData needs that update to hold up on the current set of assumptions...

A `useWorkspaces` hook that uses the workspace store and wraps `useLoadedData` to update the store would handle this.

msilva-broad replied:
hmm... ok, after our chat, I started to kick this around locally to see the benefits of going with useLoadedData variant called useAutoLoadedData that we talked about... I will have to pick this up next week after my mini-vacation:

```
export const useWorkspaces = (
  fieldsArg?: FieldsArg,
  stringAttributeMaxLength?: number
): UseWorkspacesStateResult => {

  const signal = useCancellation();
  const getData = async (): Promise<WorkspaceWrapper[]> => await workspaceProvider.list(fields, {stringAttributeMaxLength, signal});

  const [ workspaces, updateWorkspaces] = useAutoLoadedData<WorkspaceWrapper[]>(getData, {
    onError: () => { reportError('Error loading workspace list'); }
  });
  useEffect(() => {
    if (workspaces.status == 'Ready') {
      workspacesStore.set(workspaces.state);
    }
  }, [workspaces]);
  
  const fields: FieldsArg = fieldsArg || defaultFieldsArgs;
  
  return { 
    workspaces: workspacesStore.get(), 
    refresh: () => updateWorkspaces(getData), 
    loading: workspaces.status == 'Loading'
  };
};

```

nawatts commented on src/workspaces/useWorkspaces.ts:
```diff
@@ -0,0 +1,15 @@
+      useWorkspacesState,
+    }),
+};
+export const useWorkspaces = useWorkspacesComposer.make();
```
What's the purpose of `useWorkspacesComposer`? As opposed to:

```ts
export const useWorkspaces = makeUseWorkspaces({
  workspaceProvider,
  useWorkspacesState,
});
```

nawatts replied:
I'm not convinced there's value in the "is composed for Terra UI app" test... it seems like it's basically restating the function under test. So I think I'd still vote for the simpler call here.

msilva-broad replied:
there are cases where "would be really bad if this broke" is the top reason for a test... but this will be reworked based on other feedback.

nawatts replied:
Sure. But if I look at a failed test and see that it's testing the exact implementation of something I just changed, I'm probably going to update the test instead of thinking it's actually identifying an issue. If changing _anything_ in a function requires a change to a test, I'm skeptical of the usefulness of that test.

msilva-broad replied:
test reworked

nawatts commented on src/pages/EnvironmentsPage.ts:
```diff
@@ -42,7 +42,7 @@ export const EnvironmentsPage = (): ReactNode => {
     h(Environments, {
       nav: navProvider,
-      useWorkspacesState: useWorkspaces as UseWorkspacesState,
+      useWorkspacesState: useWorkspaces,
```
Here's an example of `useWorkspaces` vs `useWorkspacesState` being confusing. `useWorkspaces` has a `useWorkspacesState` dependency, but here `useWorkspaces` is used for something called `useWorkspacesState`.

msilva-broad replied:
agreed.... renamed dependency to useWorkspacesStore.... tho I am giving your other refactor comment some further thought.

nawatts replied:
I think I'd rather call the dependency `useWorkspaces`, since that's what the hook that's being passed to it is called. `useWorkspacesStore` reads like it might be `useStore(workspacesStore)`.

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- PR feedback - simplified composable pattern
- minor cleanup.

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- comment typo

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- changed stringAttributeMaxLength arg to useWorkspaces hook to only be number type.

nawatts commented on src/workspaces/useWorkspaces.composable.ts:
```diff
@@ -0,0 +1,70 @@
+  // so we can remove assumption of withErrorHandling --> --> notify() flow in terra-ui for other teams
+}
+
+export const useWorkspacesComposable = (
```
The more I think about it, I'm not sure we actually need a specialized function for this layer. This is essentially making one network request and storing the status of that request. That's also what `useLoadedData` does.

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- streamlined further to a composition based pattern where useWorkspaces adds terra-ui specific concerns to an inner useWorkspacesData hook that exposes unassuming onError and onSuccess callbacks, and leverages a new useAutoLoadedData helper hook to streamline the common pattern of auto loading on init.

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- rebase merge fix

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- rebase merge fix 2

msilva-broad commented on src/pages/ImportWorkflow/ImportWorkflow.test.ts:
```diff
@@ -21,33 +21,11 @@ type AppsContract = ReturnType<typeof Apps>;
-    // So we have to mock withWorkspaces.
-    // And since WorkspaceImporter calls withWorkspaces at the module level, it must
-    // be mocked here, not in a before... block.
-    // Thus, we also mock useWorkspaces to allow tests to control the returned workspaces.
```
none of this extra wonkiness is needed now that useWorkspaces is off on it's own module that this test file can just mock directly.

msilva-broad commented on src/workspaces/useWorkspaces.ts:
```diff
@@ -0,0 +1,45 @@
+ * @param fieldsArg
+ * @param stringAttributeMaxLength
+ */
+export const useWorkspaces = (fieldsArg?: FieldsArg, stringAttributeMaxLength?: number): UseWorkspacesStateResult => {
```
AoU would compose a similar version of useWorkspaces for use in composing Envoronments component props in their version of EnvironmentsPage.ts

msilva-broad commented on src/workspaces/useWorkspaces.ts:
```diff
@@ -0,0 +1,45 @@
+  return {
+    ...workspacesResult,
+    workspaces: workspacesStore.get(), // reflect application source-of-truth
+  };
```
A followup PR may have the hook just return the standardized `LoadedState<T>` type alongside the refresh method, but this aligns with current hook return contract for now.

nawatts commented on src/libs/ajax/loaded-data/useAutoLoadedData.ts:
```diff
@@ -0,0 +1,21 @@
+  dataCall: () => Promise<T>,
+  reloadDeps: React.DependencyList,
+  hookArgs?: UseLoadedDataArgs<T>
+): UseLoadedDataResult<T> => {
```
I think wrapping `useLoadedData` limits the benefits of this. One of the repetitive tasks that I was hoping to eliminate with this hook is removing `NoneState` from the return type.

msilva-broad replied:
I think I can manage that...

msilva-broad replied:
addressed. NoneState removed from return type

nawatts commented on src/libs/ajax/loaded-data/useLoadedData.ts:
```diff
@@ -45,40 +58,45 @@ export type UseLoadedDataResult<T> = [LoadedState<T, unknown>, (dataCall: () =>
+        });
+      }
+    },
+    [onSuccess]
```
This might break something downstream. If I remember, we wanted something more like `useState`, where the setState/updateData function is stable across renders. We intentionally moved the `onError` callback to a `useEffect` hook in #3988 so that `updateData` could be wrapped in `useCallback` without any dependencies.

msilva-broad replied:
hmmm... this was in response to a lint error since static analysis saw that onSuccess usage was now within scope of the updateData function.... but I can try the same useEffect breakout pattern you refer to...

This is a part of the React Hooks framework that I am not in love with... just how much we need to think about this stuff.

msilva-broad replied:
fixed

nawatts commented on src/workspaces/useWorkspacesData.ts:
```diff
@@ -0,0 +1,35 @@
+ * handles conversion of useAutoLoadedData mechanics to existing hook return expectations
+ * @param args - object {} must include getData method, and optionally onSuccess and/or onError
+ */
+export const useWorkspacesData = (args: UseWorkspacesDataArgs): UseWorkspacesStateResult => {
```
It's unclear what the difference between `useWorkspaces` and `useWorkspaceData` is. I think I'd rather get rid of this layer altogether to avoid that confusion.

msilva-broad replied:
if I do the follow-up refactor to just have useWorkspaces return LoadedState<T>  (or that minus the "None" state as you suggest), then I am pretty sure useWorkspacesData can go away then.

msilva-broad replied:
useWorkspacesData removed from solution.

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- more streamlining, added more reusable loadedData hooks for general use, and removing the need for useWorkspacesData hook.
- useWorkspaces now leverages new useCachedData and useLoadedDataEvents hooks to streamline composition that AoU can selectively replicate in their version of useWorkspaces hook to feed Environments component in EnvironmentsPage.ts.
- some LoadedState and related generic type cleanup.

blakery commented on src/libs/ajax/loaded-data/useLoadedData.ts:
```diff
@@ -14,13 +14,44 @@ export interface UseLoadedDataArgs<T> {
+   * Note that LoadedData object typing already allows expression of error status, convenient for consumption by
+   * visual components.  This handler is to accommodate additional side effects within the hook consuming
+   * useLoadedData hook
+   * @param state - the ready state as of when the error happened
```
Instead of `the ready state as of when the error happened`, should this be... the state before the last load? The state before the success?

Also, the example usage is for `onError`, not `onSuccess`

msilva-broad replied:
Yes, this looks like a cut-and-past from onError that wasn't fixed up.  Let me take a look.

msilva-broad replied:
fixed in earlier commit

nawatts commented on src/libs/ajax/loaded-data/useAutoLoadedData.ts:
```diff
@@ -0,0 +1,35 @@
+];
+
+/**
+ * Provides the same basic mechanics of useLoadedData hook, but will auto-load the data 'on init', or in response to reloadDeps (dependencies) argument
```
```suggestion
 * Provides the same basic mechanics of useLoadedData hook, but will auto-load the data 'on init', or in response to the reloadDeps (dependencies) argument changing
```

nawatts commented on src/libs/ajax/loaded-data/useCachedData.ts:
```diff
@@ -0,0 +1,31 @@
+import { useLoadedDataEvents } from 'src/libs/ajax/loaded-data/useLoadedData';
+import { useSettableStore } from 'src/libs/react-utils';
+
+export const useCachedData = <S, HookResult extends LoadedState<S, unknown>>(
```
Not keen on this name since, to me, "cache" implies that it's returning the value from the store instead of making the network request. But the state of the store has no effect on the network call. Maybe something along the lines of `useSyncedStore`?

msilva-broad replied:
It does return the current cached data, with a new request (or out-of-band store update) being how that cached data gets a new value.

nawatts replied:
Eh, I guess I can kind of see it. But it depends a lot on the consumer. When used with `useAutoLoadedData`, it serves as a stale while revalidate cache.

When used with `useLoadedData`, its purpose is more limited to syncing the store. A consumer could use `useLoadedData` and `useStore` directly to have the cache functionality.

This one is a little weird anyway, since a hook wrapping a hook just looks off. I'm somewhat tempted to call this machinery for creating a hook rather than a hook itself.

msilva-broad replied:
fair enough, and I've had my brain in the same space as well... suggesting useFocusedWorkspaces or something like that to indicate it's the workspaces the app is focusing on.... something to indicate it is more then just a data call wrapper... but as you said, more named towards the consumer's usage.

msilva-broad replied:
hmmm... yeah.  Tho I am trying to avoid higher-order-functions, etc... since those seem to add cognitive load....

blakery replied:
Don't all custom hooks wrap hooks, by definition?

nawatts replied:
Yes. I should have clarified, I was referring to in the call.

This pattern seems unusual:
```ts
useSomething(() => useSomethingElse())
```

msilva-broad replied:
switched useCachedData to withCachedData wrapper hook that keeps inner-hook signature for the consumer.  See latest commit.

nawatts commented on src/libs/ajax/loaded-data/useCachedData.ts:
```diff
@@ -0,0 +1,31 @@
+import { useSettableStore } from 'src/libs/react-utils';
+
+export const useCachedData = <S, HookResult extends LoadedState<S, unknown>>(
+  useHookInvoker: () => [HookResult, (dataCall: () => Promise<S>) => Promise<void>],
```
Another naming nit: `useHookInvoker` sounds like there's an extra layer: like this is a hook that gets a hook invoker. Maybe `invokeHook`? That does lose the "use" hint, but given the name makes it clear that it's invoking a hook, I think that hint is superfluous.

msilva-broad replied:
went with useLoadedDataHook and improved the jsdocs

nawatts replied:
I like it.

nawatts commented on src/libs/ajax/loaded-data/useCachedData.ts:
```diff
@@ -0,0 +1,31 @@
+    },
+  });
+
+  const finalResult =
```
One unfortunate consequence of this is that the returned data object changes every render, which has consequences for downstream `useEffect`s. Not necessarily a blocker though.

msilva-broad replied:
I've thought about this alot too... curious to pick your brain on this in a chat.

nawatts replied:
Maybe a case for `useMemo`?

blakery replied:
`useMemo` is probably a good idea, given that the potential of the data changing triggering other network calls. 
Particularly given that I'd like us to move _towards_ using effect hooks with proper dependency arrays, and _away_  from `useOnMount`.

msilva-broad replied:
agreed @blakery , tho this PR is getting long in the tooth... it makes things no worse then before (I think), so let's tackle the memo optimization as a followup PR.

nawatts commented on src/libs/ajax/loaded-data/useLoadedData.ts:
```diff
@@ -14,13 +14,44 @@ export interface UseLoadedDataArgs<T> {
+
+export type LoadedDataEvents<S, E = unknown> = Pick<UseLoadedDataArgs<S, E>, 'onSuccess' | 'onError'>;
+
+export const useLoadedDataEvents = <S, E = unknown>(loadedData: LoadedState<S, E>, events: LoadedDataEvents<S, E>) => {
```
Since there are already multiple modules for various LoadedData related hooks, can we move this to its own module?

msilva-broad replied:
agreed... I was leaning towards breaking it out myself.

msilva-broad replied:
broken out

nawatts commented on src/libs/error.ts:
```diff
@@ -3,7 +3,7 @@ import _ from 'lodash/fp';
 import { notify } from 'src/libs/notifications';
 
-export const reportError = async (title, obj) => {
+export const reportError = async (title, obj?: unknown) => {
```
Do we ever omit the `obj` argument?

msilva-broad replied:
lol, yeah, in this PR in fact... useWorkspaces hook does, with that line of code being based on the original hook's code

`      void reportError('Error loading workspace list');`

nawatts replied:
That's kind of a break with convention. Usually we include the error via `withErrorReporting`.

msilva-broad replied:
meh... just kept it as-is... error reporting/handling is it's own project to sink our teeth into...

nawatts replied:
I'll add that It's more than a break with code style convention, it's a break with existing UX patterns. Not passing the error prevents the user from being able to see information about the error.

msilva-broad replied:
yes, I TOTALLY agree.... but that's a sweater I don't want to pull on yet...

nawatts commented on src/pages/EnvironmentsPage.test.ts:
```diff
@@ -84,9 +84,12 @@ describe('Environments Page navProvider', () => {
-    const watcher = asMockedFn(mockNav.getLink);
-    expect(watcher).toBeCalledTimes(1);
-    expect(watcher).toBeCalledWith(terraNavKey('workspace-dashboard'), { name: 'myName', namespace: 'myNamespace' });
+    const getLinkWatcher = asMockedFn(mockNav.getLink);
```
Nit: It'd be nice to get rid of the inconsistency where some mocks are named with a "watcher" suffix.

This could be addressed by modifying the "arrange" section to:
```ts
const getLink = jest.fn();
const goToPath = jest.fn();
const mockNav: TerraNavLinkProvider = { getLink, goToPath };
```

msilva-broad replied:
asMockedFn adds a type check to make sure mockNav.getLink is in fact a jest mock function... but running the test will produce an actionable error if this is violated... 

Let me try this out...

nawatts commented on src/workspaces/useWorkspaces.models.ts:
```diff
@@ -0,0 +1,16 @@
+/**
+ * the expected hook return for useWorkpaces hook usage.
+ */
+export interface UseWorkspacesStateResult {
```
Some things (these types, Environments' prop) are named "UseWorkspacesState", whereas the hook is named "useWorkspaces". Can we make names more consistent by removing the "State"?

```suggestion
export interface UseWorkspacesResult {
```

msilva-broad replied:
to be honest, I am inclined to rename the hook to useWorkspacesState since it encompases the current state (cached, loading-newer-data, etc).  We've touched on this naming challenge a few times... we shorthand to useWorkspaces (useObjectX) but there would be better clarity of we named hooks as:
- useObjectXData - just a data requestor
- useObjectXState - more state complexities with ability to update/refresh, etc.

thoughts?

nawatts replied:
In my opinion, "data" and "state" are so generic that those names don't convey any meaningful distinction to me. Also, generally "use thing data" seems redundant with just "use thing".

And when we're talking about injecting these into components, like with Environments, I think we'd want the component's expected interface to be more generic. Like the component shouldn't care if there's a store involved. All it needs is something that returns the workspaces.

msilva-broad replied:
fair enough, and I've had my brain in the same space as well... suggesting useFocusedWorkspaces or something like that to indicate it's the workspaces the app is focusing on.... something to indicate it is more then just a data call wrapper... but as you said, more named towards the consumer's usage.

msilva-broad replied:
but... as we consider code reuse more and more... assuming those added app-state mechanics is probably not good either... which lands me to your suggestion...

useWorkspaces that returns UseWorkspacesResult...

LOL, See, I go there... thanks for your patience :)

msilva-broad replied:
renamed

nawatts commented on src/workspaces/useWorkspaces.test.ts:
```diff
@@ -0,0 +1,62 @@
+    ]);
+
+    // Act
+    const hookRender1 = await renderHookInAct(useWorkspaces);
```
Style note: I've been using this naming convention for `renderHook`.

```suggestion
    const { result: hookReturnRef } = await renderHookInAct(useWorkspaces);
```

msilva-broad replied:
hmmm... I kinda like having the .result.current be near the act since that lines up with cases where I want to save multiple result.current values for later assertions...
```

    // Act
    const hookRender = await renderHookInAct(useWorkspaces);
    const hookResult1 = hookRender.result.current;

    // Assert
    expect(workspaceProvider.list).toBeCalledTimes(1);
    expect(hookResult1.workspaces).toEqual([]);
```

nawatts replied:
The other style doesn't prevent that.

```
// Act
const { result: hookReturnRef } = await renderHookInAct(useWorkspaces);
const hookResult1 = hookReturnRef.current;

// Assert
expect(workspaceProvider.list).toBeCalledTimes(1);
expect(hookResult1.workspaces).toEqual([]);
```

Originally, I think we came up with that style to avoid something like:
```ts
const result = renderHook(...)
expect(result.result.current)...
```

And to clarify that the `result` field behaves like a ref. Testing Library's docs show examples with:
```ts
const { result } = renderHook(...)
```
https://testing-library.com/docs/react-testing-library/api/#renderhook

https://github.com/DataBiosphere/terra-ui/pull/3468#discussion_r1006213031

nawatts commented on src/workspaces/useWorkspaces.ts:
```diff
@@ -0,0 +1,50 @@
+  });
+
+  const hookResult: UseWorkspacesStateResult = {
+    workspaces: workspaces.state !== null ? workspaces.state : [],
```
I don't _think_ the workspaces store should ever contain null? It's initialized to an empty array.

msilva-broad replied:
LoadedState type allows null for "no value" cases.... so this is working within those mechanics to provide the default empty array []

nawatts replied:
Oh, I see.

nawatts commented:
I'm compelled to note that it would have been nice to do a preparatory PR to move `useWorkspaces` to its own module. That would have removed a lot of plumbing changes from this PR's diff.

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- PR feedback, including breaking out useLoadedDataEvents to it's own module file.
- improved jsdoc for useCachedData.

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- minor PR feedback, typos

s-rubenstein reviewed (approved):
Looks fine from a data explorer standpoint.

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- switched useCachedData to withCachedData hook wrapper function
- simplified useWorkspaces hook type names to UseWorkspaces and UseWorkspacesResult
- simplified Environments component props name to just useWorkspaces.

jdcanas reviewed (approved):
Looks good from IA file changes

msilva-broad committed:
UIE-93 Decouple data clients - useWorkspaces
- rebase and minor TS fix

cahrens commented:
Looks like @blakery didn't approve this before he headed out for vacation. I will thumb it with the caveat that I've only skimmed the code.

msilva-broad commented:
> Looks like @blakery didn't approve this before he headed out for vacation. I will thumb it with the caveat that I've only skimmed the code.

Thanks Christina.
